crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -6619,19 +6616,48 @@ impl<'db> TypeInferenceBuilder<'db> {
 
         builder = builder.add_positive(KnownClass::Bool.to_instance(self.db()));
 
+        let mut has_success = None;
+        let mut first_error = None;
+
         for pos in intersection.positive(self.db()) {
             let result = match intersection_on {
-                IntersectionOn::Left => {
-                    self.infer_binary_type_comparison(*pos, op, other, range)?
+                IntersectionOn::Left => self.infer_binary_type_comparison(*pos, op, other, range),
+                IntersectionOn::Right => self.infer_binary_type_comparison(other, op, *pos, range),
+            };
+
+            match result {
+                Ok(ty) => {
+                    has_success = Some(true);
+                    builder = builder.add_positive(ty);
                 }
-                IntersectionOn::Right => {
-                    self.infer_binary_type_comparison(other, op, *pos, range)?
+                Err(error) => {
+                    has_success = has_success.or(Some(false));
+                    first_error = first_error.or(Some(error));
                 }
-            };
-            builder = builder.add_positive(result);
+            }
         }
 
-        Ok(builder.build())
+        match has_success {
+            Some(true) | None => Ok(builder.build()),
+            Some(false) => match first_error {
+                Some(error) => Err(error),
+                None => {
+                    // This should be unreachable
+                    debug_assert!(false);
+
+                    let (left_ty, right_ty) = match intersection_on {
+                        IntersectionOn::Left => (Type::Intersection(intersection), other),
+                        IntersectionOn::Right => (other, Type::Intersection(intersection)),
+                    };
+
+                    Err(CompareUnsupportedError {
+                        op,
+                        left_ty,
+                        right_ty,
+                    })
```

Comment:
It would be great if we could get rid of this by employing a "make illegal states unrepresentable" strategy.